### ISSUE TITLE
Shader compiler is seeing the request to enable the extension in IE11

### ISSUE THREAD

**Issue by SlavaDev** [uber/luma.gl#324]
This message was in IE11,  and related with  https://github.com/uber/deck.gl/issues/966 

`deck.gl error while initializing ScatterplotLayer({id: 'scatterplotLayer'})
Error: Could not parse GLSL compiler error: Shader compilation errors
(27, 1): Unknown extension: GL_OES_standard_derivatives
(27, 48): Internal compiler error
(36, 1): Required extension GL_EXT_draw_buffers is not supported
(36, 41): Internal compiler error`

**Comment by ibgreen** [comment]
The safest approach here is probably to check for extensions on the JavaScript side and conditionally inject the code block for each extension, this way there will be nothing for the shader compiler to see if the extension is not supported.

**Comment by SlavaDev** [comment]
Hello, seems like IE still try to compile that, after update to patch 4.0.5. I have this problem 
But some parts of lessons work, exclude lessons: 05, 06, 07, 08, 16.
```
(27, 1): Unknown extension: GL_OES_standard_derivatives
(27, 48): Internal compiler error
(36, 1): Required extension GL_EXT_draw_buffers is not supported
(36, 41): Internal compiler error
```

**Comment by 1chandu** [comment]
Thanks for reporting it, we will investigate this, but in the last patch we add these lines to the shader only if the corresponding extension is enabled, so it is strange why there are still issues.

`GL_OES_standard_derivatives` is added only if `gl.getExtension('OES_standard_derivatives')` returns true, and similarly `GL_EXT_draw_buffers` is added only if   `gl.getExtension('WEBGL_draw_buffers')` returns true. As mentioned earlier due to un availability of windows machine we can't test this, if possible can you check in your app if above `gl.getExtension` calls return true?

Re-opening. @ibgreen

**Comment by SlavaDev** [comment]
@1chandu Hi, could you explain how can I test it, please?

**Comment by 1chandu** [comment]
@SlavaDev , you can print these any where in your code or one of the examples where you have access to `gl`, like in lesson-05 you can add following lines at this line : https://github.com/uber/luma.gl/blob/master/examples/lessons/05/app.js#L46
```
console.log(`OES_standard_derivatives supported: ${Boolean(gl.getExtension('OES_standard_derivatives'))}`);

console.log(`WEBGL_draw_buffers supported: ${Boolean(gl.getExtension('WEBGL_draw_buffers'))}`);
```

**Comment by SlavaDev** [comment]
@1chandu Hello, tested it in Chrome and IE-11. In the IE result: `OES_standard_derivatives supported: true` and `WEBGL_draw_buffers supported: false`, in the latest Chrome: `OES_standard_derivatives supported: false polyfill-context.js:205 WEBGL_draw_buffers supported: false`

**Comment by 1chandu** [comment]
>IE: `WEBGL_draw_buffers supported: false`

When this is false, we shouldn't be adding those instructions to the shader, I am not sure why you are getting errors related to this extension. Unfortunately I don't have access to IE system and can you debug more into this?

You need to look into `context-features.js` which has module to check for extension availability, and `getVersionDefines()` in `platform-defines.js` which conditionally adds the extensions. If you make any changes to make it work for IE and have PR I am happy to look into it.

**Comment by SlavaDev** [comment]
@1chandu Hello, I tested it on the IE-11: afterr commented

**Comment by SlavaDev** [comment]
Hello @1chandu , I tested it on the IE-11: after commented
 `/*if (hasFeature(gl, FEATURES.GLSL_DERIVATIVES)) {
      ...
  }
  if (hasFeature(gl, FEATURES.GLSL_FRAG_DATA)) {
    ...
  }*/`  in getVersionDefines() in platform-defines.js,
it works in IE-11. Tested it coupled with deck.gl (used version '4.0.5' of luma.gl, because I can't run 'master' in the IE-11 because it uses https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals  that unreadable for IE).  I had  in my case:  `Boolean(gl.getExtension('WEBGL_draw_buffers'))`  in the context-features.js  always returns "true", even if added `Boolean(gl.getExtension('Hello World'))` it returns 'true' also. Like a hot fix I suggest to add `const isIE11 = !!window.MSInputMethodContext && !!document.documentMode;` to the check of 'GLSL_DERIVATIVES' and 'FEATURES.GLSL_FRAG_DATA', but I'm not sure that it's good idea :)

**Comment by 1chandu** [comment]
>  Boolean(gl.getExtension('Hello World')) it returns 'true' also

This mean, any extension query on IE11 is going to return true, if so, this is a serious bug on IE11.

**Comment by SlavaDev** [comment]
@1chandu Hello, Will you make any changes related with it?

**Comment by 1chandu** [comment]
@SlavaDev , we would have to verify this on our side before we add any workarounds, it is hard to believe that `gl.getExtension` always returns true for any string. I will try to get back to you by end of this week, mean while if you can test, what does `getSupportedExtensions()` return?

Check: https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/getExtension

**Comment by SlavaDev** [comment]
@1chandu Hi, maybe I made something wrong before. I tested it apart from luma.gl, there are next extensions: "WEBGL_compressed_texture_s3tc", "OES_texture_float", "OES_texture_float_linear", "EXT_texture_filter_anisotropic", "OES_standard_derivatives", "ANGLE_instanced_arrays", "OES_element_index_uint", "WEBGL_debug_renderer_info" in the IE-11 (os Windows 7). I got it used this code
 `var canvas = document.getElementById('canvas');
            var gl = canvas.getContext('experimental-webgl');
            console.log(gl);
            console.log(gl.getSupportedExtensions());`

**Comment by ilyabo** [comment]
I can reproduce the issue in IE 11 on Windows 10 Pro with deck.gl@5.0.0-beta.2  and luma.gl@5.0.0-beta.1 ![image](https://user-images.githubusercontent.com/351828/34264926-b50421f4-e674-11e7-9d7c-fe3e4d730506.png)

**Comment by 1chandu** [comment]
Thanks @ilyabo for reporting this. As I mentioned in my comment above (Oct 18) do you mind making those changes to lesson-05 and paste the result here. Unfortunately I can't reproduce this locally,  I am interested to see what does `gl. getExtension` return.

**Comment by ilyabo** [comment]
Sorry for not responding right away. I tried to modify lesson-05 as you described, but unfortunately another error prevents the app to run to that place: 
![screen shot 2018-01-08 at 17 27 47](https://user-images.githubusercontent.com/351828/34681722-02b5921c-f49d-11e7-9224-ccfa0fc444a9.png)

I could do it though in my own app with a custom deck.gl layer to which I added this:

 ```
 initializeState() {
    const { gl } = this.context as any;
    console.log(`OES_standard_derivatives supported: ${Boolean(gl.getExtension('OES_standard_derivatives'))}`);
    console.log(`WEBGL_draw_buffers supported: ${Boolean(gl.getExtension('WEBGL_draw_buffers'))}`);
    ...
 }
```

This is the output I get:

![image](https://user-images.githubusercontent.com/351828/34681007-d9056bce-f49a-11e7-916e-0159ce0a24fa.png)

As you see, it returns true for `OES_standard_derivatives` but then complains that it isn't supported. Weird.

BTW, you can easily run Windows for testing apps on a Mac or Linux by using [the free Windows virtual machines from Microsoft](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) which they publish for exactly this purpose. They have images for VirtualBox which is also free.

I used this image for testing:
![image](https://user-images.githubusercontent.com/351828/34681673-e5113842-f49c-11e7-9ada-1263f2170c74.png)

**Comment by 1chandu** [comment]
Thanks @ilyabo for details.

> As you see, it returns true for OES_standard_derivatives but then complains that it isn't supported. Weird.

Yes, this is weird, I wonder if this is a known IE bug?

**Comment by ilyabo** [comment]
I couldn't find such information. I assume that it is indeed supported. How can we make sure that the extension is really enabled in the shaders?

**Comment by 1chandu** [comment]
Resolved by #444

**Comment by MarcusPenn** [comment]
This hasn't fixed the issue with GL_EXT_draw_buffers.

Depending where you make the call to the hasFeatures method with FEATURES.GLSL_FRAG_DATA you get a different result.
